### PR TITLE
fix: work with elements that have Ref's generated using the Hooks API

### PR DESCRIPTION
**Pull request by jamestalmage**
With React Hooks, ref's no longer have to be functions, but can
be simple objects with a `current` property. Instead of calling the
`ref` function, React sets `ref.current`.

Currently, the following will throw an error:

```jsx
const MyComponent = ({connectDragSource}) => {
  const ref = useRef();

  return connectDragSource(<div ref={ref}/>);
};

DragSource(type, spec, connect)(MyComponent);
```

There is ongoing work to provide a Hooks API for react-dnd, which will ultimately be preferable, but this provides an immediate fix for mixing the current API with React Hooks.

### CODE REVIEW

**Comment by RudyWeber** [comment]
Cool! I was about to submit a PR doing just that. Note that this change allows to use refs created using `React.createRef` (since React 16.3) which `useRef` uses and would close issue #998.

File: packages/react-dnd/src/utils/cloneWithRef.ts [hunk 1]
```diff
@@ -24,7 +24,11 @@ export default function cloneWithRef(
 		ref: (node: any) => {
 			newRef(node)
 			if (previousRef) {
-				previousRef(node)
+				if (typeof previousRef === 'function') {
+					previousRef(node)
+				} else {
```
Comment (RudyWeber) [inline packages/react-dnd/src/utils/cloneWithRef.ts]: Maybe
```javascript
else if (typeof previousRef === 'object' && 'current' in previousRef)
```
?
Just to be sure we are dealing with an actual ref.

Comment (jamestalmage) [reply packages/react-dnd/src/utils/cloneWithRef.ts]: Sure.

**Comment by airhorns** [comment]
This would be really great to get in if y'all had a minute or two, being able to use the new refs API would be very helpful!

**Comment by darthtrevino** [comment]
So I merged a modified PR. As a clarification, `useRef()` will still not work out but `React.createRef()` should. Let me know if there are any issues with it.

**Pull request closed without merging**